### PR TITLE
Allow inference to explore multiple instances of the same symbol

### LINKED ISSUE #31616: Type alias substitution does not work
**TypeScript Version:** 3.5.0-dev.20190525

**Search Terms:**
type substitution is incorrect

**Code**

```ts
const v = { test: { smth: 5 } };
type Field<A extends string, R> = { [K in A]: R } | {}
const f = <A extends string, B extends string, R>(x: { [K in A]: Field<B, R> } | {}): R => ({} as any);
const g = <A extends string, B extends string, R>(x: Field<A, Field<B, R>>): R => ({} as any);
const r1 = f(v); // number
const r2 = g(v); // unknown
```

**Expected behavior:**
Both should be `number`. 

**Actual behavior:**
`unknown`

**Playground Link:**
[Link](http://www.typescriptlang.org/play/#src=const%20v%20%3D%20%7B%20test%3A%20%7B%20smth%3A%205%20%7D%20%7D%3B%0D%0Atype%20Field%3CA%20extends%20string%2C%20R%3E%20%3D%20%7B%20%5BK%20in%20A%5D%3A%20R%20%7D%20%7C%20%7B%7D%0D%0Aconst%20f%20%3D%20%3CA%20extends%20string%2C%20B%20extends%20string%2C%20R%3E(x%3A%20%7B%20%5BK%20in%20A%5D%3A%20Field%3CB%2C%20R%3E%20%7D%20%7C%20%7B%7D)%3A%20R%20%3D%3E%20(%7B%7D%20as%20any)%3B%0D%0Aconst%20g%20%3D%20%3CA%20extends%20string%2C%20B%20extends%20string%2C%20R%3E(x%3A%20Field%3CA%2C%20Field%3CB%2C%20R%3E%3E)%3A%20R%20%3D%3E%20(%7B%7D%20as%20any)%3B%0D%0Aconst%20r1%20%3D%20f(v)%3B%20%2F%2F%20number%0D%0Aconst%20r2%20%3D%20g(v)%3B%20%2F%2F%20unknown)

**Related Issues:** 
Nope.

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #31616

Different instances of a type with a symbol are different instances, ergo can have different underlying structure and produce differing inferences. Stopping after we've seen a symbol once as we recur downward (but not once globally) is actually strange - it meant we'd infer to both instances in `{a: Foo<A, M>, b: Foo<B, N>}`, but to only one while checking `{a: Foo<A, Foo<B, M>>}`. The need for a limiter here makes sense (since we do produce types that can infinitely expand), but this one was excessively restrictive.

Additionally, for consistency (since it needs a similar cutoff), I've unified the depth limit constant between inference and type printout.

### CODE REVIEW

**Comment by weswigham** [comment]
@typescript-bot test this

**Comment by weswigham** [comment]
So, looking at `fp-ts` in our rwc suite, when `maxSymbolRecusionDepth` is 1, we compile in about 23 seconds. When `maxSymbolRecusionDepth` is 2 we compile in about 29 seconds. When `maxSymbolRecusionDepth` is >=3 we OOM. I'm given to believe that there is a _specific_ pattern of construction used within `fp-ts` that _happens_ to be 3 symbols deep when it triggers runaway pingponging between type inference and type creation which is why #15863 (which cutoff at depth 1) quelled it. From what I can tell, it's essentially that we're really naive when comparing `None<A>` and `FantasyMonad<"Option", A>` (and similar). When we're doing a relationship check, it's a given that we consider the thing we're comparing as true - however, in the process of structural checking we need to compare `None<(x: A) => B>` and `FantasyMonad<"Option", (x: A) => B>` - given that since we're already comparing `None<A>` and `FantasyMonad<"Option", A>`, this specialization of that should obviously be considered true given other constraints (eg, `Ternary.Maybe`), but we have nothing in place to capture this insight (and so do a generative structural comparison that only bottoms out when all limiters are triggered and only _then_ return `Ternary.Maybe`).

I have an idea for optimizing the patterns that cause the type blowup in #15443, but the first step is to actually get a test into the compiler suite of the issue. I guess I'll do that work here~

**Comment by weswigham** [comment]
I've isolated the OOM issue from `fp-ts` and pushed it to this PR (so it should now fail until I've got the fix done) - the minimum required code is actually rather small and quite regular in construction (and also stumbled upon a minor change that could be made to `fp-ts` to greatly speedup checking in the current compiler without relying on a fix here).

**Comment by weswigham** [comment]
I've re-fixed #15443 in the PR (now with a test in our suite), this time by adding an optimization into assignment checking that looks for identical types among the type parameters of the `source` and `target` and exploits those to look for a _broader_ type (usually based on a type parameter) which we may have already started to inspect (and therefore consider maybe true). With this PR in, `fp-ts` now takes around 23s to check, so about the same as before this PR, but now with a higher inference symbol depth limit (up to 5, from effectively 1).

@typescript-bot perf test this since this touches assignment checking now
@typescript-bot run dt and
@typescript-bot test this as well for good measure

**Comment by weswigham** [comment]
@typescript-bot run dt again - this update should fix those OOMs in DT with further improvements to how we calculate our max recursion depth in inference.

**Comment by fatcerberus** [comment]
> So, looking at fp-ts in our rwc suite, when maxSymbolRecusionDepth is 1, we compile in about 23 seconds. When maxSymbolRecusionDepth is 2 we compile in about 29 seconds. When maxSymbolRecusionDepth is >=3 we OOM.

It's just like `TREE(3)` :smile:

**Comment by weswigham** [comment]
Since @ahejlsberg was slightly interested in this, I've synced this PR with `master`; here's a new batch fo test runs:

@typescript-bot perf test this
@typescript-bot run dt
@typescript-bot test this
@typescript-bot user test this

**Comment by weswigham** [comment]
Perf looks good, and the changes seem good in the places with changes (sometimes it's less errors from better matching recursive structures, but in some places we issue a new error from improved inference around `flatten` or flatten-list object stuctures where we now infer both the `T` and the `T[]`, which confounds some things), but `immutable` crashes in the user suite with an OOM, so I'll have to look into why the assignability opimizations that helped `fp-ts` weren't sufficient for `immutable`.

**Comment by weswigham** [comment]
@typescript-bot perf test this
@typescript-bot run dt
@typescript-bot test this
@typescript-bot user test this

`immutable` should no longer crash; I've fixed the bug I introduced in the merge and also added some special casing for `any` in the assumptive broadening comparisons.

**Comment by weswigham** [comment]
@ahejlsberg said to punt this from 3.8 because of it's complexity, but it's still going to need a timely review to get into 3.9; so that'd be much appreciated ❤️

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -15652,6 +15659,57 @@ namespace ts {
...
                 }
 
+                // If a more _general_ version of the source and target are being compared, consider them related with assumptions
+                // eg, if { x: Q } and { x: Q, y: A } are being compared and we're about to look at { x: Q' } and { x: Q', y: A } where Q'
+                // is some specialization or subtype of Q
+                // This is difficult to detect generally, so we scan for prior comparisons of the same instantiated type, and match up matching
+                // type arguments into sets to create a canonicalization based on those matches
+                if (relation !== identityRelation && ((source.aliasSymbol && !source.aliasTypeArgumentsContainsMarker && source.aliasTypeArguments) || (getObjectFlags(source) & ObjectFlags.Reference && !!getTypeArguments(<TypeReference>source).length && !(getObjectFlags(source) & ObjectFlags.MarkerType))) &&
+                ((target.aliasSymbol && !target.aliasTypeArgumentsContainsMarker && target.aliasTypeArguments) || (getObjectFlags(target) & ObjectFlags.Reference && !!getTypeArguments(<TypeReference>target).length && !(getObjectFlags(target) & ObjectFlags.MarkerType)))) {
```
Comment (sandersn) [inline src/compiler/checker.ts]: might be worth it to extract a function for this predicate, if only to make the name obvious.

(the line length is also too long)

File: tests/baselines/reference/promiseTypeInference.errors.txt [hunk 2]
```diff
@@ -1,38 +0,0 @@
...
-    declare class Promise<T> {
-        then<U>(success?: (value: T) => Promise<U>): Promise<U>;
-    }
-    interface IPromise<T> {
-        then<U>(success?: (value: T) => IPromise<U>): IPromise<U>;
-    }
-    declare function load(name: string): Promise<string>;
-    declare function convert(s: string): IPromise<number>;
-    
-    var $$x = load("something").then(s => convert(s));
```
Comment (sandersn) [inline tests/baselines/reference/promiseTypeInference.errors.txt]: why is this error gone now? does overload resolution follow the new code path in assignability now?

Comment (weswigham) [reply tests/baselines/reference/promiseTypeInference.errors.txt]: We now successfully identify that `Promise<T>` and `IPromise<T>` are identical, and so no longer fail inference when relating their `then` signatures~

**Comment by sandersn** [comment]
Unfortunately, we never finished reviewing this PR. It is pretty old now, so I'm going to close it to reduce the number of open PRs.

**Pull request closed without merging**